controller/instance_manager_controller.go

```diff
@@ -296,140 +295,185 @@ func (imc *InstanceManagerController) syncInstanceManager(key string) (err error
 		}
 	}()
 
-	pod, err := imc.ds.GetPod(im.Name)
-	if err != nil {
-		return errors.Wrapf(err, "cannot get pod for instance manager %v", im.Name)
-	}
-
+	// Skip the cleanup when the node is suddenly down.
 	isDown, err := imc.ds.IsNodeDownOrDeleted(im.Spec.NodeID)
 	if err != nil {
 		log.WithError(err).Warnf("cannot check IsNodeDownOrDeleted(%v) when syncInstanceManager ", im.Spec.NodeID)
 	}
 	if isDown {
 		im.Status.CurrentState = longhorn.InstanceManagerStateUnknown
-		if err := imc.cleanupInstanceManager(im); err != nil {
-			return err
-		}
 		return nil
 	}
 
-	// Sync instance manager status with related pod
-	if pod == nil {
-		if im.Status.CurrentState != longhorn.InstanceManagerStateStopped {
-			im.Status.CurrentState = longhorn.InstanceManagerStateError
-		}
-	} else if pod.DeletionTimestamp != nil {
-		if im.Status.CurrentState != longhorn.InstanceManagerStateError {
-			im.Status.CurrentState = longhorn.InstanceManagerStateError
+	if err := imc.syncStatusWithPod(im); err != nil {
+		return err
+	}
+
+	if err := imc.handlePod(im); err != nil {
+		return err
+	}
+
+	if err := imc.syncInstanceManagerPDB(im); err != nil {
+		return err
+	}
+
+	if err := imc.syncInstanceManagerAPIVersion(im); err != nil {
+		return err
+	}
+
+	if err := imc.syncMonitor(im); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// syncStatusWithPod updates the InstanceManager based on the pod current phase only,
+// regardless of the InstanceManager previous status.
+func (imc *InstanceManagerController) syncStatusWithPod(im *longhorn.InstanceManager) error {
```

Comment:
It's fine to have unit tests for this function even if the pod retrieval is inside it. It's better not to move the retrieval outside of the function since there is almost no other function that requires the pod info.